pkg/certificates/certificates.go

```diff
@@ -186,18 +245,68 @@ func (c CertSignerConfig) NewAuthDB() (*AuthorizationDB, error) {
 func (c CertSignerConfig) Run(ctx context.Context, server *provider.Provider) (err error) {
 	defer mon.Task()(&ctx)(&err)
 
-	srv := NewServer(zap.L())
+	authDB, err := c.NewAuthDB()
+	if err != nil {
+		return err
+	}
+	defer func() {
+		err = errs.Combine(err, authDB.Close())
+	}()
+
+	signer, err := c.CA.Load()
+	if err != nil {
+		return err
+	}
+
+	srv := &CertificateSigner{
+		Logger:        zap.L(),
+		Signer:        signer,
+		AuthDB:        authDB,
+		MinDifficulty: uint16(c.MinDifficulty),
+	}
 	pb.RegisterCertificatesServer(server.GRPC(), srv)
 
 	return server.Run(ctx)
 }
 
 // Sign signs a valid certificate signing request's cert.
 func (c CertificateSigner) Sign(ctx context.Context, req *pb.SigningRequest) (*pb.SigningResponse, error) {
-	// lookup authtoken
-	// sign cert
-	// send response
-	return &pb.SigningResponse{}, nil
+	grpcPeer, ok := peer.FromContext(ctx)
+	if !ok {
+		// TODO: better error
+		return nil, errs.New("unable to get peer from context")
+	}
+
+	peerIdent, err := identity.PeerIdentityFromPeer(grpcPeer)
+	if err != nil {
+		return nil, err
+	}
+
+	signedPeerCA, err := c.Signer.Sign(peerIdent.CA)
+	if err != nil {
+		return nil, err
```

Comment:
Is Sign() a GRPC method?  Do we ever have any error messages that we want to hide from the client?

Reply:
It is; I'm pretty sure all error messages are intended for client consumption. We are actively preventing activation token `String` method from printing more than the first little bit of the token data for this reason. I'm not aware of any other sensitive info that might show up in error messages but you make a good point.